### DISCUSSION TITLE
On the design of generic value parameters

### DISCUSSION THREAD

**Discussion by kyouko-taiga** [hylo-lang/hylo#1212] [Language design]
As I started implementing generic (aka compile-time) value parameters (see #1211), I realized that our design has many rough edges that we should fix. I'm starting this thread to discuss them. The following are a semi-organized description of my observations so far.

### The passing convention of compile-time parameters

Run-time parameters have one of 4 passing conventions: `set`, `let`, `inout`, or `sink`. The precise meaning of these conventions is strongly related to some notion of temporality, which is a little tricky to define at compile-time.

I think it would be simpler, at least to start, to say that the passing convention of all generic value parameters is `let`. The idea is that when we write this:

```
fun foo<n: Int>() -> Int { n.copy() }
```

We're writing a family of functions that all access some global constant, distinct for each member of that family. Global constants are always `let` bindings, and therefore so are generic value parameters.

This interpretation may clash with some ideas that have been discussed for type parameters. In particular, if static `sink` subscripts are allowed. The rationale was that an expression of the form `T.property` was notionally creating an instance of `T` and then consuming it to produce some property. I don't think this model doesn't work if generic parameters have the `let` convention.

### The lifetime of compile-time parameters

If compile-time arguments are understood as global constant, then they are "immortal" and never deinitialized.

### The initialization of compile-time parameters

When compile-time arguments are initialized?

I think we should not promise any particular evaluation order, on any compile-time expression.

If we go that route, then I think one way to represent compile-time values is to have some trait that explains to the compiler how they can be serialized and then deserialized in the monomorphized form of a function, or at run-time before we call an existentialized function.

**Comment by dabrahams** [comment]
My instinct is that compile-time value parameters should act like rvalue expressions wherever they appear in expression context.  That makes them a little different to deal with than regular parameters, but at the same time, making the user call out `.copy()` (and `.move()`) seems silly, since there need be no  such operation. If we go with the "global constant" view, in addition to the static sink problem, won't there be some generic parameters that can't be returned from functions because they are immovable?

Brain slowing down; may answer the rest later…

**Reply by kyouko-taiga** [reply]
> My instinct is that compile-time value parameters should act like rvalue expressions wherever they appear in expression context.

We can make that work, I think, if we deserialize the compile-time computed representation of the value on demand.

**Reply by kyouko-taiga** [reply]
One thing to consider, though, is that it would make this code legal:

```
fun eat<T>(_ x: sink T) {}

fun foo<v: ExpensiveToDeserialize>() {
  eat(v)
  eat(v)
  eat(v)
}
```

Maybe the concern is overblown, but if `ExpensiveToDeserialize` is ... well, expensive to deserialize, we'd be hiding this cost in `foo`.

**Reply by dabrahams** [reply]
We're not deserializing at compile time?

maybe we need 

```
fun deserialize<x: T, T>() -> T { /* magic */ }
```

Or some such, to make the cost apparent?  Maybe the thing gets deserialized once automatically in each scope where it's used? I don't know, truly.

**Reply by kyouko-taiga** [reply]
> We're not deserializing at compile time?

The problem is that we need a way to teach the compile how to write in the binary the value that it computed, at compile-time. For simple things like integers we can just write literals in LLVM IR, but there isn't necessary a simple way to write any arbitrary data structure. For instance, what should we write in the binary for a particular instance of `any Regular`?

We need this representation to respect resilience boundaries, deal with run-time representations that depend on their address, and be architecture-agnostic. These constraints mean that simply capturing a snapshot of the in-memory representation of a compile-time value won't work.

So my idea was that we would have a trait à la `Codable` to represent things that can have a compile-time representation. The trait would teach the compiler how to go from the in-memory representation of a value it computed to an array of bytes that it can write in the binary. At run-time, we would deserialize these bytes to reconstruct the value.

Deserialization should be fast to make this setup work, but it's not necessarily "free", especially if we let the user write their own implementation. And sadly I think we do because we can't guess what constitutes the value of an arbitrary type.

If we go with this idea, we can choose if we want to deserialize generic value arguments once per function or once per use. If we go with the former, we'd be implementing the view that generic parameters are notionally equivalent to global constants. If we go with the latter, we'd be implementing the view that any occurrence of a generic parameter is a rvalue creating a new independent instance at run-time, with its own lifetime.

Both approaches seems to be sound and I don't know which is the best. I think I prefer the first one (global constants). It puts less constraints on parameters because it less us use immovable and/or indestructible things as generic value arguments.

**Reply by dabrahams** [reply]
I think global constants are far, far simpler to implement.  You don't even need to do this serialization thing; you can just initialize them the way other global constants would be initialized, at runtime.

**Reply by kyouko-taiga** [reply]
Ah, indeed. Expressed as a source to source transformation it would look like this:
```
// Before
fun foo<n: Int>() -> Int { n + n }

public fun main() {
  print(foo<20 + 1>())
  print(foo<90 + 7>())
}

// After
let _n1 = 20 + 1
let _n2 = 90 + 7

fun foo1() -> Int { _n1 + _n1 }
fun foo2() -> Int { _n2 + _n2 }

public fun main() {
  print(foo1())
  print(foo2())
}
```
That is even simpler than what I had in mind.

**Reply by kyouko-taiga** [reply]
But no, wait. That defeats the purpose of compile-time evaluation. We don't want to compute `90 + 7` at run-time, not even once. We can write the resulting value as a global constant, but the whole point is that it has to be already computed. I think we still need deserialization, otherwise we're just providing syntax to define generic constants (which might also be useful but is a different feature IMO).

For example:
```
// Before
fun some_formula<log: LogTable>(n: Float64) -> Float64 { n * log(n) }

public fun main() {
  let f = some_formula<LogTable(up_to: 500)>
  for let n in some_work_list {
    print(f(n))
  }
}

// After
let _logtable1 = deserialize_some_bytes() // presumably faster than re-computing a log table
...
```

**Reply by dabrahams** [reply]
I think there are two separate things here.  One is how compile-time constants will make the bridge to runtime, the other, if we choose to make them lvalues, is to constant-fold the initializations.  I fully expect that in some cases, for some types, the latter will be impossible, or at least beyond our initial capabilities.

**Reply by dabrahams** [reply]
The serialization scheme seems like a way to declaratively ensure that some types can be almost-constant-folded. As you point out, it won't always be free, and in the end the deserialization is a computation, too, so it doesn't really achieve our goal for many types where we would like it to.

**Comment by tothambrus11** [comment]
Our current plan is to use an interpreter to compute values at compile-time.
